### ISSUE TITLE
Cannot rebuild Examine Index

### ISSUE THREAD

**Issue by JarFrank** [umbraco/Umbraco-CMS#14096]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

11.3.0

### Bug summary

Hi!
Today, I updated my Umbraco 11 site to Umbraco 11.3 and discovered that my web search functionality is no longer producing any results. Typically, when faced with this type of issue, I attempt to rebuild the indexes. However, I'm unable to do so due to the following error message:

`The process is taking longer than expected, check the Umbraco log to see if there have been any errors during this operation`

I tried rebuilding the Internal, External, and two custom indexes, but none of them seem to work.

Error message:
`Error occurred executing "workItem".`
and a stack trace:
```
System.NotSupportedException: Variation "de,<null>" is not supported by the property type.
   at Umbraco.Cms.Core.Models.Property.SetValue(Object value, String culture, String segment)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.GetNestedResults(String keyPrefix, String culture, String segment, Boolean published, IDictionary`2 propertyTypeDictionary, TItem nestedContentRowValue)+MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.Handle(TSerialized deserializedPropertyValue, IProperty property, String culture, String segment, Boolean published)
   at Umbraco.Cms.Core.PropertyEditors.JsonPropertyIndexValueFactoryBase`1.GetIndexValues(IProperty property, String culture, String segment, Boolean published)
   at Umbraco.Cms.Infrastructure.Examine.BaseValueSetBuilder`1.AddPropertyValue(IProperty property, String culture, String segment, IDictionary`2 values)
   at Umbraco.Cms.Infrastructure.Examine.ContentValueSetBuilder.GetValueSetsEnumerable(IContent[] content, Dictionary`2 creatorIds, Dictionary`2 writerIds)+MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at Umbraco.Cms.Infrastructure.Examine.ContentIndexPopulator.IndexPublishedContent(Int32 contentParentId, Int32 pageIndex, Int32 pageSize, IReadOnlyList`1 indexes)
   at Umbraco.Cms.Infrastructure.Examine.ContentIndexPopulator.PopulateIndexes(IReadOnlyList`1 indexes)
   at Umbraco.Cms.Infrastructure.Examine.IndexPopulator.Populate(IIndex[] indexes)
   at Umbraco.Cms.Infrastructure.Examine.ExamineIndexRebuilder.RebuildIndex(String indexName, TimeSpan delay, CancellationToken cancellationToken)
   at Umbraco.Cms.Infrastructure.Examine.ExamineIndexRebuilder.<>c__DisplayClass9_1.<RebuildIndex>b__1()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Umbraco.Cms.Infrastructure.HostedServices.QueuedHostedService.BackgroundProcessing(CancellationToken stoppingToken)
```

### Steps to reproduce

Upgrade Umbraco 11 to 11.3

---
_This item has been added to our backlog [AB#28491](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/28491)_

**Comment by kjac** [comment]
Hi @JarFrank,

Thank you so much for reporting this. I can reproduce it - this is something we'll have to patch as soon as we can.

For the time being you can revert to the previous indexing behaviour by replacing the core index value factory for block based editors with an implementation of the default one. It sounds complicated, but fortunately it's only a few lines of code:

```cs
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.PropertyEditors;

namespace Umbraco.Cms.Web.UI.Custom;

public class MyBlockValuePropertyIndexValueFactory : DefaultPropertyIndexValueFactory, IBlockValuePropertyIndexValueFactory
{
}

public class MyBlockValuePropertyIndexValueFactoryComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder) => builder.Services.AddSingleton<IBlockValuePropertyIndexValueFactory, MyBlockValuePropertyIndexValueFactory>();
}
```

For future reference, this happens for culture variant block list and block grid properties. Culture invariant content is _not_ affected.

We'll look into releasing a patch in near future to solve this issue.

**Comment by JarFrank** [comment]
Thank you for the quick reply! I tried to implement this workaround and was able to rebuild the external index and my custom De index, but still had trouble rebuilding the internal and En indexes.

this is message `Error occurred executing "workItem".`
and this is stack trace:
```
System.NotSupportedException: Variation "de,<null>" is not supported by the property type.
   at Umbraco.Cms.Core.Models.Property.SetValue(Object value, String culture, String segment)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.GetNestedResults(String keyPrefix, String culture, String segment, Boolean published, IDictionary`2 propertyTypeDictionary, TItem nestedContentRowValue)+MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.Handle(TSerialized deserializedPropertyValue, IProperty property, String culture, String segment, Boolean published)
   at Umbraco.Cms.Core.PropertyEditors.JsonPropertyIndexValueFactoryBase`1.GetIndexValues(IProperty property, String culture, String segment, Boolean published)
   at Umbraco.Cms.Infrastructure.Examine.BaseValueSetBuilder`1.AddPropertyValue(IProperty property, String culture, String segment, IDictionary`2 values)
   at Umbraco.Cms.Infrastructure.Examine.ContentValueSetBuilder.GetValueSetsEnumerable(IContent[] content, Dictionary`2 creatorIds, Dictionary`2 writerIds)+MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at Umbraco.Cms.Infrastructure.Examine.ContentIndexPopulator.IndexAllContent(Int32 contentParentId, Int32 pageIndex, Int32 pageSize, IReadOnlyList`1 indexes)
   at Umbraco.Cms.Infrastructure.Examine.ContentIndexPopulator.PopulateIndexes(IReadOnlyList`1 indexes)
   at Umbraco.Cms.Infrastructure.Examine.IndexPopulator.Populate(IIndex[] indexes)
   at Umbraco.Cms.Infrastructure.Examine.ExamineIndexRebuilder.RebuildIndex(String indexName, TimeSpan delay, CancellationToken cancellationToken)
   at Umbraco.Cms.Infrastructure.Examine.ExamineIndexRebuilder.<>c__DisplayClass9_1.<RebuildIndex>b__1()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Umbraco.Cms.Infrastructure.HostedServices.QueuedHostedService.BackgroundProcessing(CancellationToken stoppingToken)
```

**Comment by kjac** [comment]
Hi @JarFrank,

The provided code should tackle Block List (and Block Grid) ... are you using Nested Content by any chance?

**Comment by kjac** [comment]
Hi again,

My bad, sorry. I shouldn't have assumed you used Block List.

Try this code, then:

```cs
public class MyNestedContentPropertyIndexValueFactory : DefaultPropertyIndexValueFactory, INestedContentPropertyIndexValueFactory
{
}

public class MyNestedContentPropertyIndexValueFactoryComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder) => builder.Services.AddSingleton<INestedContentPropertyIndexValueFactory, MyNestedContentPropertyIndexValueFactory>();
}
```

**Comment by JarFrank** [comment]
No worries! We actually use both, slowly migrating to BlockGrids. thank again for quick response :)

**Comment by argentini** [comment]
I have this issue as well, and I also see this exception when I save and publish a content item that has a block list property with one or more blocks in it. If the block list property is empty it does not throw. This only happens in a multi-lingual site.

**Comment by kjac** [comment]
Hi @argentini,

Thank you for reporting. It confirms our initial analysis that this does indeed "only" affect multi-lingual setups.

We're planning a patch release to fix this in the coming weeks. For now the code in [this comment](https://github.com/umbraco/Umbraco-CMS/issues/14096#issuecomment-1508425584) should revert your site back to pre-11.3 Examine behavior.

**Comment by PerplexDaniel** [comment]
@kjac I also ran into this issue in 10.5 yesterday. This error looks similar to something we encountered at some point in our package (Perplex.ContentBlocks) and the fix was to properly set the `.Variations` property on the `IPropertyType`, perhaps this is something you guys need to do as well so I thought I'll drop the snippet here (taken from [here](https://github.com/PerplexDigital/Perplex.ContentBlocks/blob/master/src/Perplex.ContentBlocks.Core/PropertyEditor/ContentBlocksValueEditor.cs#L143)):

```csharp
if (culture != null) propType.Variations |= ContentVariation.Culture;
if (segment != null) propType.Variations |= ContentVariation.Segment;
```

You may have to adjust the way you build the `propertyTypeDictionary` [here](https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.Infrastructure/PropertyEditors/NestedPropertyIndexValueFactoryBase.cs#L39) and perhaps do something like this (untested code):

```csharp
var propertyTypeDictionary =
    contentType
        .PropertyGroups
        .SelectMany(x => x.PropertyTypes!)
        .Select(propType => 
        {
            // Ensure the requested variation is allowed on the property type
            if (culture != null) propType.Variations |= ContentVariation.Culture;
            if (segment != null) propType.Variations |= ContentVariation.Segment;
            
            return propType;
        })
        .ToDictionary(x => x.Alias);
```

Again, no idea if this is the actual fix or if it something completely different but this information may be of use so I thought I'd share.

**Comment by kjac** [comment]
Thanks a lot @PerplexDaniel ... every bit of input helps. We have pinpointed the (suspected) root cause, now we "just" need to fix and test and test some more 😆

**Comment by Muttok** [comment]
Samme issue causing all of indexes to fail on friday.
Solved for now with a custom PropertyIndexValueFactoryComposer as suggested here https://github.com/umbraco/Umbraco-CMS/issues/14096#issuecomment-1508425584
That was a really frustrating friday night for a "minor" 10.4.2 => 10.5 upgrade ;)

**Comment by Zeegaan** [comment]
Fixed in https://github.com/umbraco/Umbraco-CMS/pull/14167

**Comment by nikolajlauridsen** [comment]
Just to add we've released patches for 10 and 11 with this fix that is now live on nuget 😄

**Comment by beingonlinedave** [comment]
I upgraded to 11.3.1 but I still get the issue, when trying to rebuild an index. Thanks for any help anyone can be.

System.NotSupportedException: Variation "<null>,<null>" is not supported by the property type.
   at Umbraco.Cms.Core.Models.Property.SetValue(Object value, String culture, String segment)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.GetNestedResults(String keyPrefix, String culture, String segment, Boolean published, IDictionary`2 propertyTypeDictionary, TItem nestedContentRowValue)+MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.Handle(TSerialized deserializedPropertyValue, IProperty property, String culture, String segment, Boolean published)
   at Umbraco.Cms.Core.PropertyEditors.JsonPropertyIndexValueFactoryBase`1.GetIndexValues(IProperty property, String culture, String segment, Boolean published)
   at Umbraco.Cms.Infrastructure.Examine.BaseValueSetBuilder`1.AddPropertyValue(IProperty property, String culture, String segment, IDictionary`2 values)
   at Umbraco.Cms.Infrastructure.Examine.ContentValueSetBuilder.GetValueSetsEnumerable(IContent[] content, Dictionary`2 creatorIds, Dictionary`2 writerIds)+MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at Umbraco.Cms.Infrastructure.Examine.ExamineUmbracoIndexingHandler.DeferedReIndexForContent.<>c__DisplayClass6_0.<Execute>b__0(CancellationToken cancellationToken)
   at Umbraco.Cms.Infrastructure.HostedServices.QueuedHostedService.BackgroundProcessing(CancellationToken stoppingToken)

**Comment by smarshallsay** [comment]
I can also confirm this is still an issue in 11.3.1, #14167 does not fix this

**Comment by Zeegaan** [comment]
@beingonlinedave @smarshallsay sorry to hear that this didn't fix your issue!
Can any of you reproduce this on a fresh install with reproduction steps ? 🕵️  
It would help us cover both your use case, and the one described in this issue 🙏

**Comment by nikolajlauridsen** [comment]
Hey @beingonlinedave thanks a lot for your help, we we're able to reproduce the issue and have made a fix in #14287 which is headed for the next release 🙌

**Comment by SATC-Ben** [comment]
Hi @beingonlinedave,

We updated from 11.1.0 to 11.3.1 and started getting this issue when the index builds. I have also updated it to 11.4.0 but with the same results. For context, we use the default language (en-US) but are situated in Australia with dd/mm/yyyy formats. 

[13:43:15 ERR] Error occurred executing workItem.
System.InvalidOperationException: Cannot assign value "8/31/2022 12:00:00 AM" of type "System.String" to property "toDate" expecting type "System.DateTime".
   at Umbraco.Cms.Core.Models.Property.ThrowTypeException(Object value, Type expected, String alias)
   at Umbraco.Cms.Core.Models.Property.TryConvertAssignedValue(Object value, Boolean throwOnError, Object& converted)
   at Umbraco.Cms.Core.Models.Property.SetValue(Object value, String culture, String segment)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.GetNestedResults(String keyPrefix, String culture, String segment, Boolean published, IDictionary`2 propertyTypeDictionary, TItem nestedContentRowValue, IEnumerable`1 availableCultures)+MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.Handle(TSerialized deserializedPropertyValue, IProperty property, String culture, String segment, Boolean published, IEnumerable`1 availableCultures)
   at Umbraco.Cms.Core.PropertyEditors.JsonPropertyIndexValueFactoryBase`1.GetIndexValues(IProperty property, String culture, String segment, Boolean published, IEnumerable`1 availableCultures)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.GetNestedResults(String keyPrefix, String culture, String segment, Boolean published, IDictionary`2 propertyTypeDictionary, TItem nestedContentRowValue, IEnumerable`1 availableCultures)+MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at Umbraco.Cms.Core.PropertyEditors.NestedPropertyIndexValueFactoryBase`2.Handle(TSerialized deserializedPropertyValue, IProperty property, String culture, String segment, Boolean published, IEnumerable`1 availableCultures)
   at Umbraco.Cms.Core.PropertyEditors.JsonPropertyIndexValueFactoryBase`1.GetIndexValues(IProperty property, String culture, String segment, Boolean published, IEnumerable`1 availableCultures)
   at Umbraco.Cms.Infrastructure.Examine.BaseValueSetBuilder`1.AddPropertyValue(IProperty property, String culture, String segment, IDictionary`2 values, IEnumerable`1 availableCultures)
   at Umbraco.Cms.Infrastructure.Examine.ContentValueSetBuilder.GetValueSetsEnumerable(IContent[] content, Dictionary`2 creatorIds, Dictionary`2 writerIds)+MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at Umbraco.Cms.Infrastructure.Examine.ContentIndexPopulator.IndexPublishedContent(Int32 contentParentId, Int32 pageIndex, Int32 pageSize, IReadOnlyList`1 indexes)
   at Umbraco.Cms.Infrastructure.Examine.ContentIndexPopulator.PopulateIndexes(IReadOnlyList`1 indexes)
   at Umbraco.Cms.Infrastructure.Examine.IndexPopulator.Populate(IIndex[] indexes)
   at Umbraco.Cms.Infrastructure.Examine.ExamineIndexRebuilder.RebuildIndex(String indexName, TimeSpan delay, CancellationToken cancellationToken)
   at Umbraco.Cms.Infrastructure.Examine.ExamineIndexRebuilder.<>c__DisplayClass9_1.<RebuildIndex>b__1()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Umbraco.Cms.Infrastructure.HostedServices.QueuedHostedService.BackgroundProcessing(CancellationToken stoppingToken)
